shelley/chain-and-ledger/executable-spec/test/Rules/TestLedger.hs

```diff
@@ -104,6 +112,46 @@ consumedEqualsProduced = withTests (fromIntegral numberOfTests) . property $ do
           (balance u' + d' + fees' + foldl (+) (Coin 0) rewards')
 
 
+-- | Check that a `RegPool` certificate properly adds a stake pool.
+registeredPoolIsAdded :: Property
+registeredPoolIsAdded = do
+  withTests (fromIntegral numberOfTests) . property $ do
+    tr <- forAll
+          $ traceOfLengthWithInitState @LEDGER
+                                     (fromIntegral traceLen)
+                                     mkGenesisLedgerState
+            `ofLengthAtLeast` 1
+    assertAll (addedRegPool (tr ^. traceEnv)) (sourceSignalTargets tr)
+
+ where
+
+  getTxRegPoolParams :: SourceSignalTarget LEDGER -> [PoolParams]
+  getTxRegPoolParams sst =
+    mapMaybe getPoolParams (toList (signal sst ^. body . certs))
+
+  getPoolParams :: DCert -> Maybe PoolParams
+  getPoolParams cert = case cert of
+                         RegPool pps -> Just pps
+                         _ -> Nothing
+
+  addedRegPool :: Environment LEDGER
+               -> SourceSignalTarget LEDGER
+               -> Bool
+  addedRegPool env sst = all check (getTxRegPoolParams sst)
+   where
+    check :: PoolParams -> Bool
+    check poolParams =
+      let hk = poolParams ^. poolPubKey
+          pSt = snd (target sst) ^. pstate
+          -- PoolParams are registered in pParams map
+       in M.lookup hk (pSt ^. pParams) == Just poolParams
+          -- Hashkey is registered in stPools map
+       && M.lookup hk (pSt ^. stPools . to (\(StakePools x) -> x))
+            == Just (ledgerSlot env)
+          -- Hashkey is registered in cCounters map
+       && hk ∈ M.keys (pSt ^. cCounters)
```

Comment:
These checks reflect the STS state update almost verbatim ([source](https://github.com/input-output-hk/cardano-ledger-specs/blob/53182a016429b9d2eaa380497cb6125c98a38d69/shelley/chain-and-ledger/executable-spec/src/STS/Pool.hs#L57-L59)). Part of me thinks this is almost contrived (duplicated logic between test and src), but I think we ought to check it.

Reply:
I think this is reasonable - @mgudemann what do you think?

Reply:
For `registeredPoolIsAdded`, you can extract the pool logic into the existing `TestPool.hs`, which has other properties also to be connected to the LEDGER STS.

This is the same pattern used for `rewardZeroAfterReg` and `credentialRemovedAfterDereg` in TestLedger.hs (where we put the deleg. logic in TestDeleg and then in TestLedger just connect the TestDeleg property to the LEDGER STS)

Reply:
@uroboros ah, I hadn't thought of looking for `Rules.TestPool`! I only saw that the types wouldn't fit for `Rules.TestDeleg`, because `SourceSignalTarget` didn't include `PState`. So I moved it up to `Rules.TestLedger`.

I'll relocate it.

Reply:
Moved this around in my last commit.